internal/backend/auth/authhttpmiddleware/authmiddleware.go

```diff
@@ -27,145 +28,141 @@ var Configs = configset.Set[Config]{
-			r = r.WithContext(ctxWithUserID(r.Context(), session.UserID))
+			return session.UserID, nil
 		}
 
-		next.ServeHTTP(w, r)
+		return sdktypes.InvalidUserID, nil
 	}
 }
 
-func newSetDefaultUserMiddleware(next http.Handler) http.HandlerFunc {
-	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
-		next.ServeHTTP(w, r.WithContext(ctxWithUserID(r.Context(), authusers.DefaultUser.ID())))
-	})
+func setDefaultUserMiddleware(r *http.Request) (sdktypes.UserID, *middlewareError) {
+	return authusers.DefaultUser.ID(), nil
 }
 
 func New(deps Deps) AuthMiddlewareDecorator {
 	sessions, users, tokens := deps.Sessions, deps.Users, deps.Tokens
 
-	return func(next http.Handler) http.Handler {
-		// Order matters here!
+	var mws []middlewareFn
+
+	if tokens != nil {
+		mws = append(mws, newTokensMiddleware(tokens))
+	}
 
-		// Evaluated last.
-		f := ifAuthenticated(
-			/* authenticated: */ http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
-				// make sure the user exists.
-				ctx := r.Context()
+	if sessions != nil {
+		mws = append(mws, newSessionsMiddleware(sessions))
+	}
 
-				uid := getCtxUserID(ctx)
+	if deps.Cfg.UseDefaultUser {
+		mws = append(mws, setDefaultUserMiddleware)
+	}
 
-				u, err := users.Get(authcontext.SetAuthnSystemUser(ctx), uid, "")
-				if err != nil {
-					http.Error(w, "unknown user", http.StatusUnauthorized)
-					return
+	return func(next http.Handler) http.Handler { // = AuthMiddlewareDecorator
+		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			// Process all middlewares.
+			var (
+				uid   sdktypes.UserID
+				mwErr *middlewareError
+			)
+
+			for _, mw := range mws {
+				// Stop processing middlewares if a valid user ID is found or there is an error.
+				if uid, mwErr = mw(r); uid.IsValid() || mwErr != nil {
+					break
 				}
+			}
 
-				if u.Status() != sdktypes.UserStatusActive {
-					http.Error(w, "user is not active", http.StatusUnauthorized)
-					return
-				}
+			l := deps.L
 
-				next.ServeHTTP(w, r.WithContext(authcontext.SetAuthnUser(ctx, u)))
-			}),
-			/* not authenticated: */ http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			if mwErr != nil {
+				l.Info("auth middleware error", zap.Error(errors.New(mwErr.msg)))
+				mwErr.apply(w)
+				return
+			}
+
+			// Check if authenticated.
+			if !uid.IsValid() {
+				l.Info("not authenticated")
 				http.Error(w, "unauthenticated", http.StatusUnauthorized)
-			}),
-		)
+				return
+			}
 
-		if deps.Cfg.UseDefaultUser {
-			f = ifAuthenticated(f, newSetDefaultUserMiddleware(f))
-		}
+			// Hydrate user and check if it's active.
+			l = l.With(zap.String("user_id", uid.String()))
 
-		if sessions != nil {
-			f = ifAuthenticated(f, newSessionsMiddleware(f, sessions))
-		}
+			ctx := r.Context()
 
-		// Evaluated first.
-		if tokens != nil {
-			f = ifAuthenticated(f, newTokensMiddleware(f, tokens))
-		}
+			u, err := users.Get(authcontext.SetAuthnSystemUser(ctx), uid, "")
```

Comment:
Not sure I love setting context for system user here. Need to make sure we remove it (defer?)

Reply:
it's setting it specifically for this call.